add option to force a reauth for a docker registry

tyhunt99 opened the pull request:
### What does this PR do?

Adds ability to specify reauth in the docker-registry pillar data structure to specify the reauth kwarg passed to docker-py login call. If not specified it defaults to old behavior of False. 

This makes credentials in the .docker/config.json file used by docker-py, which are possibly out of date or have been manually modified, to not interfere with credentials passed in pillar data without having to mange the config file via salt.
### What issues does this PR fix or reference?

Fixes #32829
### Previous Behavior

Retains old behavior of default reauth=False when calling docker login through docker-py
### New Behavior

Ability to specify if a reauth is done during docker login call
### Tests written?

No

Please review [Salt's Contributing Guide](https://docs.saltstack.com/en/latest/topics/development/contributing.html) for best practices.

Fixes #32829. Adds ability to specify reauth in the docker-registry pillar data structure to specify the reauth kwarg to docker-py login call. If not specified it defaults to old behaviour of False

Commit:
add option to force a reauth for a docker registry

Fixes #32829. Adds ability to specify reauth in the docker-registry pillar data structure to specify the reauth kwarg to docker-py login call. If not specified it defaults to old behaviour of False

rallytime commented:
Go Go Jenkins!

rallytime commented:
Thanks for submitting this @tyhunt99.

@ticosax Does this approach look good to you? I wanted to ping you here so you had a chance to review this change as well as to see [this comment](https://github.com/saltstack/salt/issues/32829#issuecomment-241834739) that is related to the motivation of this change. You're much more involved in this code than I am, so I wanted to make sure you had a chance to see it.

@tyhunt99 Based on the response from @ticosax and if this approach looks good, then I think getting some documentation added so users will know to use this option would be good.

tyhunt99 commented:
Jenkins was quite sluggish. 

@rallytime Do I need to worry about the failing checks? It says the lint failed but didn't actually give me any feedback as to what might be wrong.

For documentation should I just add another commit to this pr?

rallytime commented:
The ubuntu and centos failures are not related to your change, however the following pylint error is:
https://jenkins.saltstack.com/job/PR/job/salt-pr-lint-n/5931/violations/file/salt/modules/dockerng.py/

And yeah, you can add the doc change in a new commit. Any new commits will trigger a new test run, too.

Commit:
added documentation for the new reuth option in docker registry configuration

also fixed a typo in the new code where I used cred instead of creds

tyhunt99 commented:
Okay I fixed the linter issue and also added what I believe is passable documentation. If there are any issues let me know.

Also are there any guidelines as to how documentation should be written?

rallytime commented:
Thanks @tyhunt99 that looks great. Thanks for updating this!

There are some guidelines for adding docs. You can find those [here](https://docs.saltstack.com/en/latest/topics/development/conventions/documentation.html#salt-docs) for future reference.

I'll give @ticosax a little more time to respond, but from my standpoint I think this looks good. Thanks!

ticosax commented:
thx @rallytime for the ping.

tyhunt99 commented:
I see the new version was released but this PR was not in in it. When will this be merged into a release?

rallytime commented:
@tyhunt99 This PR was made against the develop branch, so as it stands it won't be in a release until `Nitrogen` comes out (the current feature release name of the develop branch). I think we could backport this to a release branch though. I'll backport it to 2016.3 so that it will be included in the 2016.3.5 and the 2016.11.1 release.

tyhunt99 commented:
Awesome thanks @rallytime! How do you tell which release develop is currently pointing to?

rallytime commented:
@tyhunt99 Develop is the branch that all feature additions go into. We branch from develop about every six months to create the new feature release. 2016.11 is the latest upcoming feature release, which will be officially announced soon. These develop feature releases are codenamed after the periodic table before we cut the date-based branch. We're currently on `Nitrogen`.

Bug fixes go into the release branches themselves, and we merge those fixes forward to all newer branches. So for example, the back-port I made just now is only against the 2016.3 branch, but once it goes in, it will be merged into 2016.11, too. 

Does that help? We also have some docs explaining this a bit more [here](https://docs.saltstack.com/en/latest/topics/development/contributing.html#which-salt-branch).

tyhunt99 commented:
@rallytime  Yes that helps. Thanks!
